src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/MetricCollector.cs

```diff
@@ -2,446 +2,540 @@
+    public Task WaitForMeasurementsAsync(int numMeasurements, TimeSpan timeout, TimeProvider? timeProvider = null)
+    {
+        _ = Throw.IfLessThan(numMeasurements, 1);
+        _ = Throw.IfLessThan(timeout.Ticks, 0);
 
-    /// <summary>
-    /// Gets a list of all observable counters registered with this metrics collector.
-    /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <returns>Read only dictionary of <see cref="MetricValuesHolder{T}"/>.</returns>
-    public IReadOnlyDictionary<string, MetricValuesHolder<T>>? GetAllObservableCounters<T>()
-        where T : struct => GetAllInstruments<T>(_allObservableCounters);
+        lock (_measurements)
+        {
+            ThrowIfDisposed();
 
-    /// <summary>
-    /// Gets a list of all observable UpDown counters registered with this metrics collector.
-    /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <returns>Read only dictionary of <see cref="MetricValuesHolder{T}"/>.</returns>
-    public IReadOnlyDictionary<string, MetricValuesHolder<T>>? GetAllObservableUpDownCounters<T>()
-        where T : struct => GetAllInstruments<T>(_allObservableUpDownCounters);
+            if (_measurements.Count >= numMeasurements)
+            {
+                return Task.CompletedTask;
+            }
 
-    /// <summary>
-    /// Gets a list of all observable Gauges registered with this metrics collector.
-    /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <returns>Read only dictionary of <see cref="MetricValuesHolder{T}"/>.</returns>
-    public IReadOnlyDictionary<string, MetricValuesHolder<T>>? GetAllObservableGauges<T>()
-        where T : struct => GetAllInstruments<T>(_allObservableGauges);
-#pragma warning restore S4049 // Properties should be preferred
+            var w = new Waiter(numMeasurements);
+            _waiters.Add(w);
+
+            return w.TaskSource.Task.WaitAsync(timeout, timeProvider ?? _timeProvider, CancellationToken.None);
+        }
+    }
 
     /// <summary>
-    /// Collects all observable instruments and records their measurements.
+    /// Scan all registered recorded instruments.
     /// </summary>
-    public void CollectObservableInstruments()
+    public void RecordObservableInstruments()
     {
-        _listener.RecordObservableInstruments();
+        if (_instrument is ObservableInstrument<T>)
+        {
+            _meterListener.RecordObservableInstruments();
+        }
     }
 
-    /// <inheritdoc />
-    public void Dispose()
+    private static AggregationType GetAggregationType(Instrument instrument)
     {
-        Dispose(true);
-        GC.SuppressFinalize(this);
+        Type genericDefinedType = instrument.GetType().GetGenericTypeDefinition();
+        AggregationType aggregationType = AggregationType.Save;
```

Comment:
Nit: I think we now have 3 enum values for only 2 possible behaviors. These are names that made the most sense to me if you wanted to use them:
Aggregate => Sum
Save => LastValue
SaveOrUpdate => LastValue

Reply:
Gone.